src/Security/TwoFactor/Provider.php

```diff
@@ -57,11 +57,13 @@ public function beginAuthentication(AuthenticationContextInterface $context): bo
             return false;
         }
 
-        if (!$user->secret) {
-            return false;
+        // Generate a secret if 2FA is enforced and the user has not yet enabled it
+        if ($this->enforceTwoFactor && !$user->secret) {
+            $user->secret = random_bytes(128);
```

Comment:
Isn't here a `$user->save();` missing?

Reply:
IDK. Can we be sure that `$context->getUser()` returns an object with a `save()` method? Also it seems to work without `save()`.

Reply:
Yes we can, it returns a `Contao\User` object.

Reply:
Added in 9f3b5df0068aea47d3958bb96a82c703618619cf.

Reply:
👍